src/server/bitops_family.cc

```diff
@@ -0,0 +1,304 @@
+// Copyright 2022, Roman Gershman.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/bitops_family.h"
+
+#include <bitset>
+
+extern "C" {
+#include "redis/object.h"
+}
+
+#include <new>
+
+#include "base/logging.h"
+#include "server/command_registry.h"
+#include "server/common.h"
+#include "server/conn_context.h"
+#include "server/engine_shard_set.h"
+#include "server/error.h"
+#include "server/tiered_storage.h"
+#include "server/transaction.h"
+#include "util/varz.h"
+
+namespace dfly {
+using namespace facade;
+
+namespace {
+static const int32_t OFFSET_FACTOR = 8;  // number of bits in byte
+
+// The following is the list of the functions that would handle the
+// commands that handle the bit operations
+void BitPos(CmdArgList args, ConnectionContext* cntx);
+void BitCount(CmdArgList args, ConnectionContext* cntx);
+void BitField(CmdArgList args, ConnectionContext* cntx);
+void BitFieldRo(CmdArgList args, ConnectionContext* cntx);
+void BitOp(CmdArgList args, ConnectionContext* cntx);
+void GetBit(CmdArgList args, ConnectionContext* cntx);
+void SetBit(CmdArgList args, ConnectionContext* cntx);
+
+OpResult<std::string> ReadValue(const OpArgs& op_args, std::string_view key);
+OpResult<bool> ReadValueBitsetAt(const OpArgs& op_args, std::string_view key, uint32_t offset);
+std::string GetString(EngineShard* shard, const PrimeValue& pv);
+bool SetBitValue(uint32_t offset, bool bit_value, std::string* entry);
+
+// ------------------------------------------------------------------------- //
+//  Bits manipulation functions
+constexpr int32_t GetBitIndex(uint32_t offset) noexcept {
+  return offset % OFFSET_FACTOR;
+}
+
+constexpr int32_t GetNormalizedBitIndex(uint32_t offset) noexcept {
+  return (OFFSET_FACTOR - 1) - GetBitIndex(offset);
+}
+
+constexpr int32_t GetByteIndex(uint32_t offset) noexcept {
+  return offset / OFFSET_FACTOR;
+}
+
+uint8_t GetByteValue(const std::string& str, uint32_t offset) {
+  return static_cast<uint8_t>(str[GetByteIndex(offset)]);
+}
+
+constexpr bool CheckBitStatus(uint8_t byte, uint32_t offset) {
+  return byte & (0x1 << offset);
+}
+
+// return true if bit is on
+bool GetBitValue(const std::string& entry, uint32_t offset) {
+  const auto byte_val{GetByteValue(entry, offset)};
+  const auto index{GetNormalizedBitIndex(offset)};
+  return CheckBitStatus(byte_val, index);
+}
+
+bool GetBitValueSafe(const std::string& entry, uint32_t offset) {
+  return ((entry.size() * OFFSET_FACTOR) > (std::size_t)offset) ? GetBitValue(entry, offset)
```

Comment:
the same - casting is redundant imho. c++ know how to compare size_t and uint32_t.

Reply:
I think that produce a warning - if not I will remove it